packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
+  /// the right edge's [isStart] is set to false. When considering vertical
+  /// ordering of boxes, the top edge is the start edge, and the bottom edge is
+  /// the end edge.
+  final bool isStart;
+
+  /// The offset from the start edge of the parent [SemanticsNode] in the
+  /// direction of the traversal.
+  final double offset;
+
+  /// The node whom this edge belongs.
+  final SemanticsNode node;
+
+  @override
+  int compareTo(_BoxEdge other) {
+    return (offset - other.offset).sign.toInt();
+  }
+}
+
+/// A group of [nodes] that are disjoint vertically or horizontally from other
+/// nodes that share the same [SemanticsNode] parent.
+///
+/// The [nodes] are sorted among each other separately from other nodes.
+class _SemanticsSortGroup extends Comparable<_SemanticsSortGroup> {
+  _SemanticsSortGroup({
+    @required this.startOffset,
+  }) : assert(startOffset != null);
+
+  /// The offset from the start edge of the parent [SemanticsNode] in the
+  /// direction of the traversal.
+  ///
+  /// This value is equal to the [_BoxEdge.offset] of the first node in the
+  /// [nodes] list being considered.
+  final double startOffset;
+
+  /// The nodes that are sorted among each other.
+  final List<SemanticsNode> nodes = <SemanticsNode>[];
+
+  @override
+  int compareTo(_SemanticsSortGroup other) {
+    return (startOffset - other.startOffset).sign.toInt();
+  }
+
+  /// Sorts this group assuming that [nodes] belong to the same vertical group.
+  ///
+  /// This method breaks up this group into horizontal [_SemanticsSortGroup]s
+  /// then sorts them using [sortedWithinKnot].
+  List<SemanticsNode> sortedWithinVerticalGroup() {
+    final List<_BoxEdge> edges = <_BoxEdge>[];
+    for (SemanticsNode child in nodes) {
+      edges.add(new _BoxEdge(
+        isStart: true,
+        offset: _pointInParentCoordinates(child, child.rect.topLeft).dx,
+        node: child,
+      ));
+      edges.add(new _BoxEdge(
+        isStart: false,
+        offset: _pointInParentCoordinates(child, child.rect.bottomRight).dx,
+        node: child,
+      ));
+    }
+    edges.sort();
+
+    final List<_SemanticsSortGroup> horizontalGroups = <_SemanticsSortGroup>[];
+    _SemanticsSortGroup group;
+    int depth = 0;
+    for (_BoxEdge edge in edges) {
+      if (edge.isStart) {
+        depth += 1;
+        group ??= new _SemanticsSortGroup(
+          startOffset: edge.offset,
+        );
+        group.nodes.add(edge.node);
+      } else {
+        depth -= 1;
+      }
+      if (depth == 0) {
+        horizontalGroups.add(group);
+        group = null;
+      }
+    }
+    horizontalGroups.sort();
+
+    final List<SemanticsNode> result = <SemanticsNode>[];
+    for (_SemanticsSortGroup group in horizontalGroups) {
+      final List<SemanticsNode> sortedKnotNodes = group.sortedWithinKnot();
+      result.addAll(sortedKnotNodes);
+    }
+    return result;
+  }
+
+  /// Sorts [nodes] where nodes intersect both vertically and horizontally.
```

Comment:
Slight confusion: Doesn't this method also get called for lists of length one that don't have overlap per definition?

Reply:
Yeah, it's one of those "Are all elements of an empty set blue?" kind of situation. It's just a special case. But I made it clearer in the docs.